### ISSUE TITLE
[Bug]: NetTCP connection doesn't send RST packet when timeout

### ISSUE THREAD

**Issue by mkopacka** [CoreWCF/CoreWCF#1345]
### Duplicate ?

- [X] I have searched issues/discussions and did not find other issues/discussions reporting this bug.

### Product version

1.5.1

### Describe expected behavior

When server (service) timeouts (disconnects / closes connection) then RST packet should be sent to the client so it knows about being disconnected.

### Describe actual behavior

When server timeouts then no RST packet is sent so client thinks it is still connected. but if it tries to get any data from the server, it timeouts (after client timeout). 

I have tried to create server and client in NET6 using CoreWCF and also in Framework 4.7.2 to compare behaviour. 

**Framework 4.7.2 behaviour:**

- Server starts, client connects - no problem 
- I set timeout of server to 10 seconds. Timeout of client to 5 seconds. 
- If I do nothing for 10 seconds, server timeouts and informs client about it
    - Client disconnects
    - It is visible in Wireshark

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/cfa3dd43-a19c-4d8f-a695-e407c7d2cb19)

Events on server side:

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/1e0336b0-cb48-4f65-a5f4-b55a152ac666)

**NET6 behaviour:**
- Server starts, client connects - no problem
- Timeouts are again - 10 seconds at server side, 5 seconds at client side
- If I do nothing for 10 seconds, I see that server closes connection, but nothing is sent to the client and therefore connection looks active, even though, it is not. 
    - If client tries to call something from service, it doesn't reach it because connection is off and timeouts in 5 seconds as not reachable. 

Server side events:
![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/cb09a928-3eee-46ed-9563-a1806bb14a61)

But nothing is sent to the client and therefore connection looks still on:

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/083b305d-7123-43cd-b88c-b9ec22ac8944)

But, if I try to call method from client again, it doesn't reach it. And timeouts in 5 seconds (client timeout)

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/fa49f365-b43f-44b1-977f-42cdcd504997)
![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/37557467-bf6f-4f05-82ba-7d687d4eaab9)

We can see in the server logs that it didn't reach it at all so connection was closed. Client just didn't know:
![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/c4233edd-0fd5-47d7-9901-9ef36ff02bf4)

### Which binding

NetTcp

### security

Transport

### Which .NET version

.NET 6

### Which os platform

Windows

### Code snippet used to reproduce the issue

```c#
// WCF Service

[ServiceContract]
public interface ITestWcfService
{
    [OperationContract]
    string GetData(int value);
}

[ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession, ConcurrencyMode = ConcurrencyMode.Multiple)]
public class TestWcfService : ITestWcfService
{
    public TestWcfService() //only instantiated by WCF
    {
        Console.WriteLine("Constructor - NET6");
        OperationContext.Current.InstanceContext.Faulted += Channel_Faulted;
        OperationContext.Current.InstanceContext.Closed += Channel_Closed;
        OperationContext.Current.InstanceContext.Opened += InstanceContext_Opened;
    }

    public string GetData(int value)
    {
        Console.WriteLine("GetData - NET6");
        try
        {
            return $"{value}";
        }
        catch (Exception ex)
        {
            throw;
        }
    }

    private void InstanceContext_Opened(object sender, EventArgs e)
    {
        Console.WriteLine("Opened - NET6");
    }

    private void Channel_Faulted(object sender, EventArgs e)
    {
        Console.WriteLine("Faulted - NET6");
    }

    private void Channel_Closed(object sender, EventArgs e)
    {
        Console.WriteLine("Closed - NET6");
    }
}

// Server:

private static WebApplication _app;
public Form1()
{
    InitializeComponent();

    var builder = WebApplication.CreateBuilder();
    builder.WebHost.UseNetTcp(11999);
    builder.Services.AddServiceModelServices();
    builder.Services.AddServiceModelMetadata();

    _app = builder.Build();

    var binding = new NetTcpBinding(SecurityMode.Transport)
    {
        CloseTimeout = TimeSpan.FromSeconds(10),
        OpenTimeout = TimeSpan.FromSeconds(10),
        ReceiveTimeout = new TimeSpan(0, 0, 0, 10),
        SendTimeout = TimeSpan.FromSeconds(10)
    };
    
    _app.UseServiceModel(serviceBuilder =>
    {
        serviceBuilder.AddService<TestWcfService>(options =>
        {
            options.DebugBehavior.IncludeExceptionDetailInFaults = true;
        });

        serviceBuilder.AddServiceEndpoint<TestWcfService, ITestWcfService>(binding, "net.tcp://localhost:11999/TestWcfService2");
    });

    _app.StartAsync().Wait();
}

// Client:

private ITestWcfService client;
public Form1()
{
    InitializeComponent();

    IClientChannel channel = null;
    var binding = new NetTcpBinding()
    {
        CloseTimeout = TimeSpan.FromSeconds(5),
        OpenTimeout = TimeSpan.FromSeconds(5),
        ReceiveTimeout = new TimeSpan(0, 0, 0, 5),
        SendTimeout = TimeSpan.FromSeconds(5),
    };

    var factory = new ChannelFactory<ITestWcfService>(binding, new EndpointAddress("net.tcp://localhost:11999/TestWcfService2"));
    factory.Open();
    try
    {
        client = factory.CreateChannel();
    }
    finally
    {
    }
}

private void button1_Click(object sender, EventArgs e)
{
    client.GetData(4);
}
```

**Comment by mconnew** [comment]
> If I do nothing for 10 seconds, I see that server closes connection, but nothing is sent to the client and therefore connection looks active, even though, it is not.

What do you mean by server closes connection? Do you mean `OperationContext.Current.InstanceContext.Closed` is called to notify you but nothing happens on the network? Or do you mean that a FIN packet is sent (as opposed to RST)?  

What happens if you attach event handlers to OperationContext.Current.Channel? Does the Closed event handler fire?  

If you call `OperationContext.Current.Channel.Abort()` does it send a RST?

**Comment by mkopacka** [comment]
> > If I do nothing for 10 seconds, I see that server closes connection, but nothing is sent to the client and therefore connection looks active, even though, it is not.
> 
> What do you mean by server closes connection? Do you mean `OperationContext.Current.InstanceContext.Closed` is called to notify you but nothing happens on the network? Or do you mean that a FIN packet is sent (as opposed to RST)?

Yes, exactly (first option) - `OperationContext.Current.InstanceContext.Closed` is executed but nothing happens on the network - neither RST nor FIN packet is sent. 

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/46d9eb3a-736e-4803-9c47-d0a51a868dc9)

> 
> What happens if you attach event handlers to OperationContext.Current.Channel? Does the Closed event handler fire?

I am not sure here. If you mean at service side (CoreWCF) then `OperationContext.Current.Channel` is not available. It would be in System.ServiceModel which is eventually only at client side.

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/f85411d7-4e22-4709-9422-86a65050e189)

> 
> If you call `OperationContext.Current.Channel.Abort()` does it send a RST?

And here same answer as above, `OperationContext.Current.Channel.` is not available. I tried to call  `OperationContext.Current.InstanceContext.Abort()` as test now but it crashes on exception. 

![image](https://github.com/CoreWCF/CoreWCF/assets/48550785/3ed78abb-e4c8-4e82-856b-ba355fdbc440)

Here is a zip with simple examples to easier/faster reproduction. It contains servers and clients in NET6 and framework472. 
[WcfTest.zip](https://github.com/CoreWCF/CoreWCF/files/14491979/WcfTest.zip)

I tried to connect to these servers also from Windows Telnet client. In both cases it doesn't instantiate the service class but in case of framework server, telnet gets disconnected after a while. However in case of NET6 server, it stays connected for eternity.

**Comment by JunoGonzalezfisglobalcom** [comment]
Hello,
We are getting a very similar issue after we migrated our wcf services to core wcf.
We replicate by calling a singleton service on the migrated core wcf web services, wait for the CoreWCF.Channels.ServiceChannel.SessionIdleManager to abort the channel when the ReceiveTimeout (https://learn.microsoft.com/en-us/dotnet/framework/wcf/feature-details/configuring-timeout-values-on-a-binding) from the NetTcpBinding is reached, then we try to call the same service again and for the client, the channel seems to be valid but when the server tries to send the response back to the client we get error below server side and the client service request times out.
 CoreWCF.CommunicationObjectAbortedException: The communication object, CoreWCF.Channels.ServerFramingDuplexSessionChannel, cannot be used for communication because it has been Aborted.
   at CoreWCF.Channels.CommunicationObject.ThrowIfDisposedOrNotOpen()
   at CoreWCF.Channels.OutputChannel.SendAsync(Message message, CancellationToken token)
   at CoreWCF.Dispatcher.DuplexChannelBinder.DuplexRequestContext.OnReplyAsync(Message message, CancellationToken token)
   at CoreWCF.Channels.RequestContextBase.ReplyAsync(Message message, CancellationToken token)
   at CoreWCF.Dispatcher.ImmutableDispatchRuntime.ReplyAsync(MessageRpc rpc)
I have checked on IServiceBehavior we use on the server side that the channel is closed and tried to do something so the client realizes the channel is not valid and triggers the channel repair policy client side but nothing has worked so far,
				public object AfterReceiveRequest(ref Message request, IClientChannel channel, InstanceContext instanceContext) {
					if (channel.State == CommunicationState.Closed || channel.State == CommunicationState.Closing || channel.State == CommunicationState.Faulted)
						throw new System.ServiceModel.CommunicationException();
I have pointed the client to the legacy wcf web services and for the same scenario it fails trying to send the request (at System.ServiceModel.Channels.StreamConnection.Write, the socket connection is closed) after the channel is closed for inactivity (ReceiveTimeout) on the server side, because a CommunicationException is thrown client side the repair policy is triggered, the channel gets repaired and the service request goes thru.
For the migrated core wcf web services we are getting timeouts on the web app side and CommunicationObjectAbortedException server side.
Is this related to core wcf not sending the RST or FIN when the channel is aborted by CoreWCF.Channels.ServiceChannel.SessionIdleManager because of inactivity (NetTcpBinding ReceiveTimeout)?
Any recommendation to workaround/fix this issue?
Thanks,
Juno

**Comment by mirek-kopacka** [comment]
I'd like to add some additional information to this. This time I only started a server and connected Telnet client to it. In this way, WcfTestService is not instantiated, so there are not any events. Reason of this check was to see netstat and ports state. 

1. So, started WCF Service - everything as expected, port is listening
![image](https://github.com/CoreWCF/CoreWCF/assets/144328427/cc115fb6-12ed-4ee9-a193-aef6a2b83579)

2. Connected telnet client to it
![image](https://github.com/CoreWCF/CoreWCF/assets/144328427/f3659b71-5f56-4bf4-b51e-682c3dfe0b16)
![image](https://github.com/CoreWCF/CoreWCF/assets/144328427/21ece5d5-bfc0-4671-b1a4-d44222aeaf7d)

3. And now I waited for timeouts again. It doesn't come, which means that port is Established for as long as client don!'t initiate to close connection. So I can connect more and more
![image](https://github.com/CoreWCF/CoreWCF/assets/144328427/2381a96a-4b48-4243-afb3-3931b85d25f0)

That basically leads to possibility of Denial of Service (DoS) attack and it actually prevent us from releasing the product. 
For that reason, I find it concerning and as high severity bug.

**Comment by mconnew** [comment]
I believe I found the missing line of code. ServerFramingDuplexSessionChannel.ReturnConnectionIfNecessary has an abort argument. In .NET Framework, the sessionReader was responsible for handling connection reuse, but in CoreWCF it's an aspnetcore middleware component that handles it. I believe it just needs:
```c#
            if (abort)
            {
                Connection.Abort();
            }
```
Adding at the top of the method. I'm going to write a test to validate the client goes into the expected state then should have a PR available pretty quickly after that.

**Comment by mconnew** [comment]
BTW, I believe a DoS attack isn't possible with this issue because it requires an authenticated connection in the first place. If a malicious client has valid credentials to a service, there are potentially a lot worse things you need to worry about. Having said that, I'm going to release a fix for this with some urgency. I'll keep this thread updated.

**Comment by mconnew** [comment]
@mirek-kopacka, I initially misread what you wrote, I'm making another change to address that. It's a change I've been wanting to do for a while, it's a pattern I created for named pipes where there's an options class to specify connection specific parameters, as the Binding isn't known until too late. The default connection timeout is 30 seconds on .NET Framework. If you want to change it, you would do something like this:
```c#
builder.UseNetTcp(options =>
{
    options.Listen("net.tcp://localhost:11999" listenOptions =>
    {
        listenOptions.ChannelInitializationTimeout = TimeSpan.FromSeconds(10);
    }
});
```
If you are happy with the default of 30 seconds, the current api's will still work.

**Comment by mkopacka** [comment]
> I believe I found the missing line of code. ServerFramingDuplexSessionChannel.ReturnConnectionIfNecessary has an abort argument. In .NET Framework, the sessionReader was responsible for handling connection reuse, but in CoreWCF it's an aspnetcore middleware component that handles it. I believe it just needs:
> 
> ```cs
>             if (abort)
>             {
>                 Connection.Abort();
>             }
> ```
> 
> Adding at the top of the method. I'm going to write a test to validate the client goes into the expected state then should have a PR available pretty quickly after that.

Thank you @mconnew for such quick fix! 

> BTW, I believe a DoS attack isn't possible with this issue because it requires an authenticated connection in the first place. If a malicious client has valid credentials to a service, there are potentially a lot worse things you need to worry about. Having said that, I'm going to release a fix for this with some urgency. I'll keep this thread updated.

Yes, you are right, it would be (most likely) my least problem in such case



> @mirek-kopacka, I initially misread what you wrote, I'm making another change to address that. It's a change I've been wanting to do for a while, it's a pattern I created for named pipes where there's an options class to specify connection specific parameters, as the Binding isn't known until too late. The default connection timeout is 30 seconds on .NET Framework. If you want to change it, you would do something like this:
> 
> ```cs
> builder.UseNetTcp(options =>
> {
>     options.Listen("net.tcp://localhost:11999" listenOptions =>
>     {
>         listenOptions.ChannelInitializationTimeout = TimeSpan.FromSeconds(10);
>     }
> });
> ```
> 
> If you are happy with the default of 30 seconds, the current api's will still work.

Noted, thank you for a tip!

**Comment by rikardhansson** [comment]
@mconnew Thank you for investigating!

> BTW, I believe a DoS attack isn't possible with this issue because it requires an authenticated connection in the first place. If a malicious client has valid credentials to a service, there are potentially a lot worse things you need to worry about. Having said that, I'm going to release a fix for this with some urgency. I'll keep this thread updated.

A generic TCP client can open a connection that stays in Established until closed by the client, according to what I see with _netstat_. .NET Framework times out and closes on server side, but I don't get a timeout on CoreWCF / .NET 6.  

No credentials involved, so this might be a risk for DoS attack. Client code:

```
using System.Net;
using System.Net.Sockets;

var address = IPAddress.Parse("127.0.0.1");
var ipEndPoint = new IPEndPoint(address, 11999);
using var client = new Socket(ipEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
await client.ConnectAsync(ipEndPoint);
await Task.Delay(TimeSpan.FromMinutes(20));
```

**Comment by mconnew** [comment]
I've just released v1.4.2 and v1.5.2 fixing both of the timeout connection closing problems. I'm going to close this issue even though a large/better fix (configurable timeouts) is in progress. The released fixes are the minimum change required to fix the problem.  

@mirek-kopacka, in future it's generally best to report potential security issues privately. We have a published security policy detailing how to do this [here](https://github.com/CoreWCF/CoreWCF/security/policy). It's important to report it privately so that there isn't a publicly known vulnerability without any fixes that could be exploited to harm companies or individuals. There's a good write up about why to do responsible disclosure in the [first few paragraphs of this blog post](https://snyk.io/blog/responsible-disclosure/).

**Comment by mirek-kopacka** [comment]
> I've just released v1.4.2 and v1.5.2 fixing both of the timeout connection closing problems. I'm going to close this issue even though a large/better fix (configurable timeouts) is in progress. The released fixes are the minimum change required to fix the problem.
> 
> @mirek-kopacka, in future it's generally best to report potential security issues privately. We have a published security policy detailing how to do this [here](https://github.com/CoreWCF/CoreWCF/security/policy). It's important to report it privately so that there isn't a publicly known vulnerability without any fixes that could be exploited to harm companies or individuals. There's a good write up about why to do responsible disclosure in the [first few paragraphs of this blog post](https://snyk.io/blog/responsible-disclosure/).

Thank you for really quick fix! Gonna test it now! 

And noted, I even didn't think about that possibility and I am sorry for that. If will have "luck" in the future, I'll do it privately for sure!
